src/compilers/v1/compiler_v1.cpp

```diff
@@ -0,0 +1,507 @@
+    using unique_ptr = std::unique_ptr<clc_object, void(*)(clc_object*)>;
+
+private:
+     unique_ptr m_Object;
+
+public:
+    ProgramBinaryV1(unique_ptr obj);
+    clc_object *GetRaw() const { return m_Object.get(); }
+
+    // Inherited via ProgramBinary
+    virtual ~ProgramBinaryV1() = default;
+    virtual bool Parse(Logger const *logger) final;
+    virtual size_t GetBinarySize() const final;
+    virtual const void *GetBinary() const final;
+};
+
+class CompiledDxilV1 : public CompiledDxil
+{
+public:
+    using unique_ptr = std::unique_ptr<clc_dxil_object, void(*)(clc_dxil_object*)>;
+
+private:
+    unique_ptr m_Object;
+
+public:
+    CompiledDxilV1(ProgramBinaryV1 const& parent, unique_ptr obj);
+    clc_dxil_object *GetRaw() const { return m_Object.get(); }
+
+    // Inherited via CompiledDxil
+    virtual ~CompiledDxilV1() = default;
+    virtual size_t GetBinarySize() const final;
+    virtual const void *GetBinary() const final;
+    virtual void *GetBinary() final;
+};
+
+static clc_logger ConvertLogger(Logger const& logger)
+{
+    auto log = [](void *ctx, const char *msg) { static_cast<Logger*>(ctx)->Log(msg); };
+    clc_logger ret;
+    ret.error = log;
+    ret.warning = log;
+    ret.priv = (void*)&logger;
+    return ret;
+}
+
+#if 0
+#endif
+
+CompilerV1::CompilerV1(XPlatHelpers::unique_module compiler)
+    : m_Compiler(std::move(compiler))
+{
+#define GET_FUNC(Member, api) Member = m_Compiler.proc_address<decltype(&api)>(#api)
+    GET_FUNC(CreateContext, clc_context_new);
+    GET_FUNC(SerializeContext, clc_context_serialize);
+    GET_FUNC(DeserializeContext, clc_context_deserialize);
+    GET_FUNC(FreeSerializedContext, clc_context_free_serialized);
+    GET_FUNC(FreeContext, clc_free_context);
+    GET_FUNC(CompileImpl, clc_compile);
+    GET_FUNC(LinkImpl, clc_link);
+    GET_FUNC(FreeSpirv, clc_free_object);
+    GET_FUNC(GetKernelImpl, clc_to_dxil);
+    GET_FUNC(FreeDxil, clc_free_dxil_object);
+    GET_FUNC(GetCompilerVersion, clc_compiler_get_version);
+#undef GET_FUNC
+
+    if (!CreateContext ||
+        !FreeContext ||
+        !CompileImpl ||
+        !LinkImpl ||
+        !FreeSpirv ||
+        !GetKernelImpl ||
+        !FreeDxil)
+        throw std::runtime_error("Failed to load required compiler entrypoints");
+
+    m_Context = decltype(m_Context)(nullptr, FreeContext);
+}
+
+CompilerV1 *CompilerV1::Instance()
+{
+    return static_cast<CompilerV1*>(g_Platform->GetCompiler());
+}
+
+bool CompilerV1::Initialize(ShaderCache &cache)
+{
+    if (m_Context)
+        return true;
+
+    std::lock_guard lock(m_InitializationLock);
+    if (m_Context)
+        return true;
+
+    // {1B9DC5F4-545A-4356-98D3-B4C0062E6253}
+    static const GUID ClcContextKey =
+    { 0x1b9dc5f4, 0x545a, 0x4356, { 0x98, 0xd3, 0xb4, 0xc0, 0x6, 0x2e, 0x62, 0x53 } };
+
+    if (DeserializeContext)
+    {
+        if (auto CachedContext = cache.Find(&ClcContextKey, sizeof(ClcContextKey));
+            CachedContext.first)

```

Comment:
huh?

Reply:
C++17 "if initializer" syntax. Check out https://en.cppreference.com/w/cpp/language/if "If Statements with Initializer" section. For something as simple as a null check, older "contextually convertible to bool" for variable declaration technically works, but I find this much clearer.